docs/configuration.md
---
comment: 
proofedDate: noProof
supportsVersion: v 1.0.0 SA
---

# YAML config file template

> Note, in production you must use only one `store` method, i.e. `file`, `oci`, or `aws`.

```yaml
base:
  qredoAPI: https://play-api.qredo.network/api/v1/p
  pin: 0
autoApproval:
  enabled: false
  retryIntervalMaxSec: 300
  retryIntervalSec: 5
websocket:
  qredoWebsocket: wss://play-api.qredo.network/api/v1/p/coreclient/feed
  reconnectTimeoutSec: 300
  reconnectIntervalSec: 5
  pingPeriodSec: 5
  pongWaitSec: 10
  writeWaitSec: 10
  readBufferSize: 512
  writeBufferSize: 1024
http:
  addr: 0.0.0.0:8007
  CORSAllowOrigins:
    - '*'
  logAllRequests: false
  TLS:
    enabled: true
    certFile: tls/domain.crt
    keyFile: tls/domain.key
logging:
  format: text
  level: debug
loadBalancing:
  enable: false
  onLockErrorTimeoutMs: 300
  actionIDExpirationSec: 6
  redis:
    host: localhost
    port: 6379
    password: ""
    db: 0
store:
  type: file
  file: /volume/ccstore.db
  oci:
    compartment: ocid1.tenancy.oc1...
    vault: ocid1.vault.oc1...
    secretEncryptionKey: ocid1.key.oc1...
    configSecret: signing_agent_config
  aws:
    region: aws-region-...
    configSecret: secrets_manager_secret...
```

## Base

- **qredoAPI:** the url of the api you want to use
- **pin:** the pin number to use to provide a zero knowledge proof token for communication with the partner api

## Auto approval
- **enabled:** activate the automatic approval of every transaction that is received
- **retryIntervalMaxSec:** the maximum time in which the Signing Agent retries to approve an action. After that it’s considered as a failure
- **retryIntervalSec:** the interval in which the Signing Agent is attempting to approve an action. It will retry until the retryIntervalMaxSec is reached

## Websocket
- **qredoWebsocket:** the url of the websocket feed you want to use
- **reconnectTimeoutSec:** the reconnect timeout in seconds
- **reconnectIntervalSec:** the reconnect interval in seconds
- **pingPeriodSec:** the ping period for the ping handler in seconds
- **pongWaitSec:** the pong wait for the pong handler in seconds
- **writeWaitSec:** the write wait in seconds
- **readBufferSize:** the websocket upgrader read buffer size in bytes
- **writeBufferSize:** the websocket upgrader write buffer size in bytes

## HTTP

- **addr:** the address and port the service runs on [the bind address and port the build in api endpoints]
- **CORSAllowOrigins:** the value the the Access-Control-Allow-Origin of the responses of the build in api
- **logAllRequests:** log all incoming requests to the build in api
- **TLS**
  - **enabled:** wether or not you want to enable tls on the server side
  - **certFile:** path to the cert file you want to use
  - **keyFile:** path to the key file you want to use


## Logging

- **format:** the format of the logging, ex. text or json
- **level:** the level of logging, ex. debug, info, warn, error, default is debug

## Load balancing

- **enable:** enables the load-balancing logic
- **onLockErrorTimeoutMs:** on lock timeout in milliseconds
- **actionIDExpirationSec:** expiration of actionID variable in Redis in seconds
- **redis:**
  - **host:** Redis host
  - **port:** Redis port
  - **password:** Redis password
  - **db:** Redis database to be selected after connecting to the server

## Store

- **type:** the type of store to use to store the private key information for the Signing Agent, ex. file, oci, aws
- **file:** the path to the storage file when file store is used
- **oci:** the oracle cloud configuration to store the private keys in an oracle vault
  - **compartment:** the OCID where the vault and encryption key reside
  - **vault:** the OCID of the vault where the secret will be stored
  - **secretEncryptionKey:** the encryption key used for both the secret and the data inside the secret
  - **configSecret:** the name of secret that will be used to store the data
- **aws:** the amazon cloud configuration to store the private keys in amazon secrets manager
  - **region:** the AWS region where the secret is stored
  - **configSecret:** the name of the AWS Secrets Manager secret containing the encrypted data


CHANGELOG.md
# Changelog

## v1.0.0 (2022-12-13)


config/config.go
package config

import (
	"os"

	"github.com/pkg/errors"
	"gopkg.in/yaml.v2"
)

// swagger:model ConfigResponse
type Config struct {
	Base          Base            `yaml:"base" json:"base"`
	HTTP          HttpSettings    `yaml:"http" json:"http"`
	Logging       Logging         `yaml:"logging" json:"logging"`
	LoadBalancing LoadBalancing   `yaml:"loadBalancing" json:"loadBalancing"`
	Store         Store           `yaml:"store" json:"store"`
	AutoApprove   AutoApprove     `yaml:"autoApproval" json:"autoApproval"`
	Websocket     WebSocketConfig `yaml:"websocket" json:"websocket"`
}

type Base struct {
	// The pin number to use to provide a zero knowledge proof token for communication with the Partner API.
	// example: 123456
	PIN int `yaml:"pin" json:"pin"`

	// The URL of the Qredo API.
	// example: https://sandbox-api.qredo.network
	QredoAPI string `yaml:"qredoAPI" json:"qredoAPI"`
}

type TLSConfig struct {
	// Enable TLS for the internal HTTP server.
	// example: true
	Enabled bool `yaml:"enabled" json:"enabled"`

	// The cert file to use for the TLS server.
	// example: tls/domain.crt
	CertFile string `yaml:"certFile" json:"certFile"`

	// The key file to use for the TLS server.
	// example: tls/domain.key
	KeyFile string `yaml:"keyFile" json:"keyFile"`
}

type AutoApprove struct {
	// Activate the automatic approval of every transaction that is received.
	// example: true
	Enabled bool `yaml:"enabled" json:"enabled"`

	// The maximum time over which the Signing Agent retries action approval. After this period, it’s considered as a failure.
	// example: 300
	RetryIntervalMax int `yaml:"retryIntervalMaxSec" json:"retryIntervalMaxSec"`

	// The interval over which the Signing Agent attempts to approve an action. It will cycle this interval until the `retryIntervalMaxSec` is reached.
	// example: 5
	RetryInterval int `yaml:"retryIntervalSec" json:"retryIntervalSec"`
}

type WebSocketConfig struct {
	// The URL of the Qredo WebSocket feed.
	// example: wss://sandbox-api.qredo.network/api/v1/p/coreclient/feed
	QredoWebsocket string `yaml:"qredoWebsocket" json:"qredoWebsocket"`

	// The reconnect timeout in seconds.
	// example: 300
	ReconnectTimeOut int `yaml:"reconnectTimeoutSec" json:"reconnectTimeoutSec"`

	// The reconnect interval in seconds.
	// example: 5
	ReconnectInterval int `yaml:"reconnectIntervalSec" json:"reconnectIntervalSec"`

	// The ping period for the ping handler in seconds.
	// example: 5
	PingPeriod int `yaml:"pingPeriodSec" json:"pingPeriodSec"`

	// The pong wait for the pong handler in seconds.
	// example: 10
	PongWait int `yaml:"pongWaitSec" json:"pongWaitSec"`

	// The write wait in seconds.
	// example: 10
	WriteWait int `yaml:"writeWaitSec" json:"writeWaitSec"`

	// The WebSocket upgrader read buffer size in bytes.
	// example: 512
	ReadBufferSize int `yaml:"readBufferSize" json:"readBufferSize"`

	// The WebSocket upgrader write buffer size in bytes.
	// example: 1024
	WriteBufferSize int `yaml:"writeBufferSize" json:"writeBufferSize"`
}

type Store struct {
	// The type of store to use to store the private key information for the Signing Agent.
	// enum: file,oci,aws
	// example: file
	Type string `default:"file" yaml:"type" json:"type"`

	// The path to the storage file when `file` store is used.
	// example: /volume/ccstore.db
	FileConfig string `yaml:"file" json:"file"`

	OciConfig OciConfig `yaml:"oci" json:"oci"`
	AwsConfig AWSConfig `yaml:"aws" json:"aws"`
}

// OciConfig-based Signing Agent config: used when Base `store` `type` is `oci`.
type OciConfig struct {
	// The OCID where the vault and encryption key reside.
	// example: ocid1.tenancy.oc1...
	Compartment string `yaml:"compartment" json:"compartment"`

	// The OCID of the vault where the secret will be stored.
	// example: ocid1.vault.oc1...
	Vault string `yaml:"vault" json:"vault"`

	// The encryption key used for both the secret and the data inside the secret.
	// example: ocid1.key.oc1...
	SecretEncryptionKey string `yaml:"secretEncryptionKey" json:"secretEncryptionKey"`

	// The name of secret that will be used to store the data.
	// example: automated_approver_config
	ConfigSecret string `yaml:"configSecret" json:"configSecret"`
}

// AWSConfig-based Signing Agent config: used when Base `store` `type` is `aws`.
type AWSConfig struct {
	// The AWS region where the secret is stored.
	// example: eu-west-3
	Region string `yaml:"region" json:"region"`

	// The name of the AWS Secrets Manager secret containing the encrypted data.
	// example: secrets_manager_secret
	SecretName string `yaml:"configSecret" json:"configSecret"`
}

type HttpSettings struct {
	// The address and port the service runs on [the bind address and port the build in API endpoints].
	// example: 0.0.0.0:8007
	Addr string `yaml:"addr" json:"addr"`

	// The value of the Access-Control-Allow-Origin of the responses of the build in API.
	// example: '*'
	CORSAllowOrigins []string `yaml:"CORSAllowOrigins" json:"CORSAllowOrigins"`

	// Log all incoming requests to the build in API.
	// example: true
	LogAllRequests bool `yaml:"logAllRequests" json:"logAllRequests"`

	TLS TLSConfig `yaml:"TLS" json:"TLS"`
}

type Logging struct {
	// Output format of the log.
	// enum: text,json
	// example: json
	Format string `yaml:"format" json:"format"`

	// Log level to be output.
	// enum: info,warn,error,debug
	// example: debug
	Level string `yaml:"level" json:"level"`
}

type LoadBalancing struct {
	// Enables the load-balancing logic.
	// example: true
	Enable bool `yaml:"enable" json:"enable"`

	// The on lock timeout in milliseconds.
	// example: 300
	OnLockErrorTimeOutMs int `yaml:"onLockErrorTimeoutMs" json:"onLockErrorTimeoutMs"`

	// The expiration of actionID variable in Redis in seconds.
	// example: 6
	ActionIDExpirationSec int         `yaml:"actionIDExpirationSec" json:"actionIDExpirationSec"`
	RedisConfig           RedisConfig `yaml:"redis" json:"redis"`
}

// RedisConfig-based Signing Agent config: used when LoadBalancing `enable` is `true`.
type RedisConfig struct {
	// The Redis host.
	// example: redis
	Host string `yaml:"host" json:"host"`

	// The Redis port.
	// example: 6379
	Port int `yaml:"port" json:"port"`

	// The Redis password.
	// example: just a password
	Password string `yaml:"password" json:"password"`

	// Redis database to be selected after connecting to the server.
	// example: 0
	DB int `yaml:"db" json:"db"`
}

// Default creates configuration with default values.
func (c *Config) Default() {
	c.HTTP = HttpSettings{
		Addr:             "127.0.0.1:8007",
		CORSAllowOrigins: []string{"*"},
		TLS: TLSConfig{
			Enabled: false,
		},
	}

	c.Base.PIN = 0
	c.Base.QredoAPI = "https://sandbox-api.qredo.network/api/v1/p"
	c.AutoApprove = AutoApprove{
		Enabled:          false,
		RetryIntervalMax: 300,
		RetryInterval:    5,
	}
	c.Websocket = WebSocketConfig{
		ReconnectTimeOut:  300,
		ReconnectInterval: 5,
		QredoWebsocket:    "wss://sandbox-api.qredo.network/api/v1/p/coreclient/feed",
		PingPeriod:        5,
		PongWait:          10,
		WriteWait:         10,
		ReadBufferSize:    512,
		WriteBufferSize:   1024,
	}
	c.Logging.Level = "info"
	c.Logging.Format = "json"
	c.Store.Type = "file"
	c.Store.FileConfig = "ccstore.db"
	c.LoadBalancing = LoadBalancing{
		Enable:                false,
		OnLockErrorTimeOutMs:  300,
		ActionIDExpirationSec: 6,
		RedisConfig: RedisConfig{
			Host:     "redis",
			Port:     6379,
			Password: "",
			DB:       0,
		},
	}
}

// Load reads and parses yaml config.
func (c *Config) Load(fileName string) error {
	f, err := os.ReadFile(fileName)
	if err != nil {
		return errors.Wrap(err, "read config file")
	}

	c.Default()
	if err := yaml.Unmarshal(f, c); err != nil {
		return errors.Wrap(err, "parse config file")
	}

	return nil
}

// Save saves yaml config.
func (c *Config) Save(fileName string) error {
	b, err := yaml.Marshal(c)
	if err != nil {
		return err
	}
	return os.WriteFile(fileName, b, 0600)
}
